### ISSUE TITLE
Support returning a status from an async port as an alternate queue full behavior

### ISSUE THREAD

**Issue by celskeggs** [nasa/fpp#555]
Currently there is no easy way for the sender of a message on an async port to observe whether the queue is full.

Let's look at adding a queue full behavior that returns a status instead of either dropping or invoking a behavior on the receiver. For example, `OK` if the message was successfully placed on the queue, and `DROPPED` if the message was dropped.

**Comment by bocchino** [comment]
@timcanham @LeStarch This seems like a good idea to me. Thoughts?

**Comment by bocchino** [comment]
I'm not sure this would be a specified queue full behavior at the receiver end. I think it would be a change to the port invocation interface. Any port type that does not have a return value in the FPP model (and so could be used in an asynchronous context) would have an implicit status return. Invoking the port would cause the status to be produced. If the port is connected to a sync port at the other end, it would always return OK. But if the port at the other end is async, it could return something else like QUEUE_FULL.

**Comment by celskeggs** [comment]
I don't think you want to impact _every_ output port that doesn't return a value. You'd either end up with lots of code sprinkled with `FW_ASSERT(status == OK);` or you'd end up with lots of code that simply ignores the status returned by a port, simply because there are many uses for synchronous ports without return values. It would introduce many opportunities for error if a port written to assume a synchronous input port (that always returns OK) was connected to an asynchronous input port (that might not always return OK).

Here's an alternative proposal: you could have a particular return type for a port that _can_ be used in an asynchronous context.

```
port BufferTrySend(ref buf: Fw.Buffer) -> Status

# Example in a component that wants to asynchronously implement the interface:
async input port try_send: BufferTrySend

# The declaration:
void try_send_handler(FwIndexType portNum, Fw::Buffer &buf) override;

# Example in a component that wants to synchronously implement the interface:
sync input port try send: BufferTrySend

# The declaration:
Status try_send_handler(FwIndexType portNum, Fw::Buffer &buf) override;
```

This would not require a major change to how we handle port calls without return values in general. It would just open the door to returning a status from one particular kind of async port. This status would be returned by the autocoded port implementation when it enqueues or fails to enqueue the message.

**Comment by bocchino** [comment]
It might be OK to require a check or a cast to `(void)`. The cast to `(void)` would be not because the output port is assumed to be connected to a sync input port, but because there's nothing to do on the sender end if the message is dropped.

If we wanted to be more rigorous in the model, we could allow an output port instance _P_ to specify that it must be connected to a sync input port; in this case the code generator would omit the status return for the port invocation function for _P_. However, this would limit the contexts in which the output port could be used.

If an output port _P_ has port type _T_, and _T_ returns data, then we know that _P_ cannot be connected to an async input port. Otherwise I don't think we should use the port type _T_ to determine whether the output invocation function for _P_ returns a status. The status return depends on the kind of connection (sync or async) and not on the port type. Also, it's a common pattern to use the same port type in different contexts (sync and async) in different components.

**Comment by bocchino** [comment]
Also, the status is not the return type of the input port handler -- an async input handler never gets called unless the message went through the queue, and it always has return type `void`. The status is the return type of the function that's called to invoke the output port, on the thread of the sending component.

**Comment by celskeggs** [comment]
I think we're talking past each other. Let's schedule a meeting to discuss this.

**Comment by bocchino** [comment]
Sounds good.

**Comment by bocchino** [comment]
Here is what we decided after discussion. Assume the following:

* A port `P`.
* A component `A` with instance `pOut` of port `P`.
* An instance `a` of component `A`.
* A component `B` with instance `pIn` of port `P`.
* An instance `b` of component `B`.
* A connection `a.pOut -> b.pIn`

1. If `P` has a return type, then the connection must be synchronous. So there is no port invocation status to return. The behavior of the generated code should be the same as it is now.
2. Otherwise, the connection may be synchronous or asynchronous, and the asynchronous case may produce errors. In this case we will provide optional syntax in the FPP model letting the port instance `A.pOut` opt in to error handling.
    a. If the opt-in syntax is not present (default behavior), then the behavior is the same as it is today: no status is reported from the invocation.
    b. If the opt-in syntax is present, then the generated C++ invocation function will have a return value, which reports the status. The status is one of SYNC, ASYNC_SUCCESS, ASYNC_DROP, ASYNC_HOOK.

This way one can see in the FPP model if any instance `a` that does not handle invocation errors is connected to an instance `b` that can produce invocation errors. One can write a checker against the model to produce a warning or error in this case.

**Comment by celskeggs** [comment]
I think we didn't entirely reach consensus on whether information on SYNC vs. ASYNC should be exposed. I think we were going to circle back about the use case for that distinction; there were concerns that it would expose information that should be hidden. Otherwise, I think this is an accurate description of what we discussed.
